### ISSUE TITLE
Hang on file upload after period of inactivity

### ISSUE THREAD

**Issue by jonathanswenson** [awslabs/aws-crt-java#665]
### Describe the bug

I'm attempting to use the CRT s3 client to upload files to s3 when I don't know their full size at time of upload. 

Uploads / Downloads are successful while I use the application, but after some time passes between requests (seems like a few hours is required), the system will hang and I'll begin to receive errors. 

```
java.lang.IllegalStateException: The service request was not made within 10 seconds of outputStream being invoked. Make sure to invoke the service request BEFORE invoking outputStream if your caller is single-threaded.
```

In the past I've noticed that this is often due to invalid request parameters, so if I catch (and ignore that error) and wait on the futures, (either the upload or download future), which typically will throw the exception, it will just hang for ever. 

### Expected Behavior

I expect that after some time has passed, the s3client / transfer manager will continue to allow for for uploads rather than hanging forever / failing to make a request. 

I guessing that I may be missing some health check settings or need to change the OS keepalive settings, 
but would expect that there would at least be an error message somewhere indicating that the connection has failed rather than an indefinite hang. 

### Current Behavior

Waiting on an upload or download future will hang (seemingly forever) and attempting to create a outputStream body will fail after the default 10 second timeout with the following exception: 
```

java.lang.IllegalStateException: The service request was not made within 10 seconds of outputStream being invoked. Make sure to invoke the service request BEFORE invoking outputStream if your caller is single-threaded. at 
software.amazon.awssdk.core.async.BlockingOutputStreamAsyncRequestBody.waitForSubscriptionIfNeeded(BlockingOutputStreamAsyncRequestBody.java:93) at 
software.amazon.awssdk.core.async.BlockingOutputStreamAsyncRequestBody.outputStream(BlockingOutputStreamAsyncRequestBody.java:68) at
... my code
```

### Reproduction Steps

The upload code looks roughly like this (kotlin), let me know and I can probably reproduce with java. 
```
// initialization
val s3ClientBuilder = S3AsyncClient.crtBuilder()
    .credentialsProvider(ProfileCredentialsProvider.create("SandboxAdmin"))

val s3Client = s3ClientBuilder.build()

val transferManager = S3TransferManager.builder().s3Client(client).build()
```

Executed per "request" using a transfermanager / s3 client that is built once on initialization
```
val body = AsyncRequestBody.forBlockingOutputStream(null)
val req = UploadRequest.builder()
    .requestBody(body)
    .putObjectRequest { it.bucket(bucket).key(key) }
    .build()

val upload = transferManager.upload(req).completionFuture()

body.outputStream().use {
  // upload to stream
}

// wait for upload to complete
upload.join()
```

I also will download files from s3 (again per request using the pre-initialized transfer manager / s3 client)
```
val downloadRequest = DownloadRequest.builder()
    .getObjectRequest { it.bucket(bucket).key(key) }
    .responseTransformer(AsyncResponseTransformer.toBlockingInputStream())
    .build()

val downloadResponse = transferManager
    .download(downloadRequest)
    .completionFuture()

return downloadResponse.join().result()
```

### Possible Solution

My best guess is that this is some sort of keepalive problem and can be worked around by changing the OS level keepalives (or perhaps using the httpConfiguration to set up a healthcheck properly)

I have also tried setting up a health check on the s3client, but that also didn't fix the issue for me. 
```
.httpConfiguration(S3CrtHttpConfiguration.builder()
    .connectionHealthConfiguration(S3CrtConnectionHealthConfiguration.builder()
        .minimumThroughputInBps(0)
        .minimumThroughputTimeout(Duration.ofSeconds(30))
        .build()
    )
    .build()
)
```

### Additional Information/Context

Seemingly related to:
+ https://github.com/awslabs/aws-c-s3/issues/210
+ https://github.com/aws/aws-sdk-cpp/issues/2107

Only successful reproduction is running in a container based off of `public.ecr.aws/docker/library/eclipse-temurin:17-jdk-jammy` running in ECS

Keepalive settings in the OS
```
# cat /proc/sys/net/ipv4/tcp_keepalive_time
7200
# cat /proc/sys/net/ipv4/tcp_keepalive_intvl 
75
# cat /proc/sys/net/ipv4/tcp_keepalive_probes
9
```

This also appears to happen on an x86 build, also running in ECS. 

### aws-crt-java version used

0.25.0

### Java version used

Open JDK 17

### Operating System and version

Linux hostname 5.10.186-179.751.amzn2.aarch64 #1 SMP Tue Aug 1 20:51:46 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux

**Comment by jonathanswenson** [comment]
I managed to reproduce with a relatively simple test. At a high level it opens up a s3 client, and then in a loop:
+ runs an upload, 
+ sleeps for some amount of time that doubles each iteration.

Dependencies: 
```
  implementation(platform("software.amazon.awssdk:bom:2.20.131"))
  implementation("software.amazon.awssdk:sso")
  implementation("software.amazon.awssdk.crt:aws-crt:0.25.0")
  implementation("software.amazon.awssdk:s3-transfer-manager")
  implementation("software.amazon.awssdk:s3")
```

```
import org.junit.jupiter.api.Test;
import software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider;
import software.amazon.awssdk.core.async.AsyncRequestBody;
import software.amazon.awssdk.core.async.BlockingOutputStreamAsyncRequestBody;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3AsyncClient;
import software.amazon.awssdk.services.s3.S3CrtAsyncClientBuilder;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.transfer.s3.S3TransferManager;
import software.amazon.awssdk.transfer.s3.model.CompletedUpload;
import software.amazon.awssdk.transfer.s3.model.UploadRequest;

import java.io.OutputStream;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class S3UploadTest {
    private void upload(S3TransferManager transferManager, String bucket, String key) throws Exception {
        BlockingOutputStreamAsyncRequestBody body = AsyncRequestBody.forBlockingOutputStream(null);
        PutObjectRequest putRequest = PutObjectRequest.builder().bucket(bucket).key(key).build();

        UploadRequest req = UploadRequest.builder().requestBody(body).putObjectRequest(putRequest).build();
        CompletableFuture<CompletedUpload> uploadFuture = transferManager.upload(req).completionFuture();

        try (OutputStream stream = body.outputStream()) {
            stream.write("Hello World!".getBytes());
        } catch (IllegalStateException ise) {
            // dead
            System.out.println("failed upload to " + bucket + "/" + key);
            ise.printStackTrace();
            // wait for upload to complete
            try {
                CompletedUpload completedUpload = uploadFuture.get(30, TimeUnit.SECONDS);
                // don't expect to ever get here.
            } catch (TimeoutException te) {
                System.out.println("upload timed out");
                te.printStackTrace();
            }
        }

        // if we got this far, we at least successfully created the output stream and likely will succeed.
        CompletedUpload completedUpload = uploadFuture.join();

        System.out.println("completed upload to " + bucket + "/" + key);
    }

    @Test
    public void testLongUpload() throws Exception {
        String bucketName = "upload-timeout-test";
        S3CrtAsyncClientBuilder s3ClientBuilder = S3AsyncClient.crtBuilder()
                .credentialsProvider(ProfileCredentialsProvider.create("SandboxAdmin"))
                .region(Region.US_WEST_2);

        try (S3AsyncClient s3Client = s3ClientBuilder.build()) {
            try (S3TransferManager transferManager = S3TransferManager.builder().s3Client(s3Client).build()) {

                long startTime = System.currentTimeMillis();
                long lastExecutionTime = startTime;
                long waitTime = 30 * 1000; // start at 30 seconds
                while (true) {
                    long currentExecutionStart = System.currentTimeMillis();
                    String key = UUID.randomUUID().toString();
                    long sinceStart = currentExecutionStart - startTime;
                    long sinceLastExecution = currentExecutionStart - lastExecutionTime;
                    System.out.println("Uploading to " + bucketName + "/" + key + " " + sinceStart + "ms since start, " + sinceLastExecution + "ms since last execution");

                    upload(transferManager, bucketName, key);
                    long uploadCompleteTime = System.currentTimeMillis();
                    lastExecutionTime = System.currentTimeMillis();
                    System.out.println("Sleeping for " + waitTime + "ms");
                    Thread.sleep(waitTime);
                    long sleepCompleteTime = System.currentTimeMillis();
                    System.out.println("Sleep complete in " + (sleepCompleteTime - uploadCompleteTime) + "ms");

                    waitTime *= 2;
                }
            }
        }
    }
}

```

it failed at the iteration where we had waited 480,000ms (8 minutes) and a total time of 933842ms (15 minutes) since start. 

I ran this on my M1 Mac with keep alive settings of: 
```
❯ sysctl -A | grep 'net.inet.tcp.*keep'
net.inet.tcp.keepidle: 7200000
net.inet.tcp.keepintvl: 75000
net.inet.tcp.keepinit: 75000
net.inet.tcp.keepcnt: 8
net.inet.tcp.always_keepalive: 0
```

<details><summary> Error output / stacktraces </summary>

```
failed upload to <bucket_name>/ee561b71-c770-491b-8927-bf727a0f412d
java.lang.IllegalStateException: The service request was not made within 10 seconds of outputStream being invoked. Make sure to invoke the service request BEFORE invoking outputStream if your caller is single-threaded.
	at software.amazon.awssdk.core.async.BlockingOutputStreamAsyncRequestBody.waitForSubscriptionIfNeeded(BlockingOutputStreamAsyncRequestBody.java:93)
	at software.amazon.awssdk.core.async.BlockingOutputStreamAsyncRequestBody.outputStream(BlockingOutputStreamAsyncRequestBody.java:68)
	at com.acme.S3UploadTest.upload(S3UploadTest.java:29)
	at com.acme.S3UploadTest.testLongUpload(S3UploadTest.java:71)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
upload timed out
java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1960)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2095)
	at com.acme.S3UploadTest.upload(S3UploadTest.java:37)
	at com.acme.S3UploadTest.testLongUpload(S3UploadTest.java:71)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
``` 

</details>

**Comment by jonathanswenson** [comment]
I've now run it a few times and seem to get a failure every time at around 15 minutes on my mac. 

I hooked up the trace logger (via `Log.initLoggingToFile(Log.LogLevel.Trace, "/tmp/s3-log");`) and am not seeing anything I would really say stands out but might be telling to someone who knows what's going on: 

For the first 5 requests I get output that looks like this (omitted the `[dns]` and `[event-loop]` output which is very noisy)

```
[INFO] [2023-08-24T03:39:58Z] [000000016d307000] [S3Client] - id=0x103d25f10 Initiating making of meta request
[DEBUG] [2023-08-24T03:39:58Z] [000000016d307000] [S3MetaRequest] - id=0x121407f40 Created new Auto-Ranged Put Meta Request.
[DEBUG] [2023-08-24T03:39:58Z] [000000016d307000] [channel-bootstrap] - id=0x6000038d3640: acquiring bootstrap reference
[INFO] [2023-08-24T03:39:58Z] [000000016d307000] [connection-manager] - id=0x121407ba0: Successfully created
[DEBUG] [2023-08-24T03:39:58Z] [000000016d307000] [S3Endpoint] - id=0x600001d17120: Created connection manager 0x121407ba0 for endpoint
[INFO] [2023-08-24T03:39:58Z] [000000016d307000] [S3Client] - id=0x103d25f10: Created meta request 0x121407f40
[DEBUG] [2023-08-24T03:39:58Z] [000000017129f000] [task-scheduler] - id=0x103d260e0: Scheduling s3_client_process_work_task task for immediate execution
[DEBUG] [2023-08-24T03:39:58Z] [000000017129f000] [task-scheduler] - id=0x103d260e0: Running s3_client_process_work_task task with <Running> status
[DEBUG] [2023-08-24T03:39:58Z] [000000017129f000] [S3Client] - id=0x103d25f10 s_s3_client_process_work_default - Moving relevant synced_data into threaded_data.
[DEBUG] [2023-08-24T03:39:58Z] [000000017129f000] [S3Client] - id=0x103d25f10 s_s3_client_process_work_default - Processing any new meta requests.
[DEBUG] [2023-08-24T03:39:58Z] [000000017129f000] [S3Client] - id=0x103d25f10 Updating meta requests.
[DEBUG] [2023-08-24T03:39:58Z] [000000017129f000] [S3Client] - id=0x103d25f10 Updating connections, assigning requests where possible.
[INFO] [2023-08-24T03:39:58Z] [000000017129f000] [S3ClientStats] - id=0x103d25f10 Requests-in-flight(approx/exact):17/1  Requests-preparing:17  Requests-queued:0  Requests-network(get/put/default/total):0/0/0/0  Requests-streaming-waiting:0  Requests-streaming:0  Endpoints(in-table/allocated):1/1
[DEBUG] [2023-08-24T03:39:58Z] [0000000171a47000] [task-scheduler] - id=0x60000381c7f0: Scheduling s3_meta_request_prepare_request_task task for immediate execution
[DEBUG] [2023-08-24T03:39:58Z] [0000000171a47000] [task-scheduler] - id=0x60000381c7f0: Running s3_meta_request_prepare_request_task task with <Running> status
[DEBUG] [2023-08-24T03:39:58Z] [0000000171a47000] [S3MetaRequest] - id=0x121407f40: Prepared request 0x600002fb0200 for part 0
[TRACE] [2023-08-24T03:39:58Z] [0000000171a47000] [S3MetaRequest] - id=0x121407f40 Created signable 0x6000033b8140 for request 0x600002fb0200 with message 0x6000033ba3f0
[DEBUG] [2023-08-24T03:39:58Z] [0000000171a47000] [aws-c-common] - s_aws_jni_get_thread_env returned detached, attaching
[INFO] [2023-08-24T03:39:58Z] [0000000171a47000] [AuthSigning] - (id=0x6000033b8140) Signing successfully built canonical request for algorithm SigV4, with contents 
POST
/15501153-26f0-48e1-928d-265c755eb876
uploads=
amz-sdk-invocation-id:46039903-c9b5-ccb7-a4ae-ced37dd61c0b
amz-sdk-request:attempt=1; max=1
content-type:application/octet-stream
host:<bucket>.s3.us-west-2.amazonaws.com
x-amz-checksum-algorithm:CRC32
x-amz-content-sha256:UNSIGNED-PAYLOAD
x-amz-date:20230824T033958Z
x-amz-security-token:<omitted>

```

The last one (the one that fails) I get output that looks like this (before I kill the process a few minutes later)
```
[INFO] [2023-08-24T03:47:59Z] [000000016d307000] [S3Client] - id=0x103d25f10 Initiating making of meta request
[DEBUG] [2023-08-24T03:47:59Z] [000000016d307000] [S3MetaRequest] - id=0x11e2178e0 Created new Auto-Ranged Put Meta Request.
[DEBUG] [2023-08-24T03:47:59Z] [000000016d307000] [channel-bootstrap] - id=0x6000038d3640: acquiring bootstrap reference
[INFO] [2023-08-24T03:47:59Z] [000000016d307000] [connection-manager] - id=0x11e217e00: Successfully created
[DEBUG] [2023-08-24T03:47:59Z] [000000016d307000] [S3Endpoint] - id=0x600001deca80: Created connection manager 0x11e217e00 for endpoint
[INFO] [2023-08-24T03:47:59Z] [000000016d307000] [S3Client] - id=0x103d25f10: Created meta request 0x11e2178e0
[DEBUG] [2023-08-24T03:47:59Z] [000000017129f000] [task-scheduler] - id=0x103d260e0: Scheduling s3_client_process_work_task task for immediate execution
[DEBUG] [2023-08-24T03:47:59Z] [000000017129f000] [task-scheduler] - id=0x103d260e0: Running s3_client_process_work_task task with <Running> status
[DEBUG] [2023-08-24T03:47:59Z] [000000017129f000] [S3Client] - id=0x103d25f10 s_s3_client_process_work_default - Moving relevant synced_data into threaded_data.
[DEBUG] [2023-08-24T03:47:59Z] [000000017129f000] [S3Client] - id=0x103d25f10 s_s3_client_process_work_default - Processing any new meta requests.
[DEBUG] [2023-08-24T03:47:59Z] [000000017129f000] [S3Client] - id=0x103d25f10 Updating meta requests.
[DEBUG] [2023-08-24T03:47:59Z] [000000017129f000] [S3Client] - id=0x103d25f10 Updating connections, assigning requests where possible.
[INFO] [2023-08-24T03:47:59Z] [000000017129f000] [S3ClientStats] - id=0x103d25f10 Requests-in-flight(approx/exact):20/0  Requests-preparing:20  Requests-queued:0  Requests-network(get/put/default/total):0/0/0/0  Requests-streaming-waiting:0  Requests-streaming:0  Endpoints(in-table/allocated):1/1
[WARN] [2023-08-24T03:55:23Z] [000000016df4f000] [JavaCrtGeneral] - Not all native threads were successfully joined during gentle shutdown.  Memory may be leaked.
```

**Comment by jonathanswenson** [comment]
Was also able to reproduce using an arm64 container based on `public.ecr.aws/docker/library/eclipse-temurin:17-jdk-jammy` running in docker with the same keep alive settings. 

```
root@4648ebd7c611:/test# cat /proc/sys/net/ipv4/tcp_keepalive_time
7200
root@4648ebd7c611:/test# cat /proc/sys/net/ipv4/tcp_keepalive_intvl 
75
root@4648ebd7c611:/test# cat /proc/sys/net/ipv4/tcp_keepalive_probes
9
```

**Comment by TingDaoK** [comment]
Thanks for reporting the issue. 

We allocated the bug, it's an internal counter issue. 

To work around the bug for quickest way, you can add a content-length to your request. It will bypass the bug.

I am currently working on a fix, will update the thread once the fix is done.

**Comment by jonathanswenson** [comment]
I cannot add content length to the request (it is not known at the time of initiating the request due to generating a stream as part of the the workload that initiates the request)

Unless I’m mistake the crt library is the only system that currently supports Java streams with an unknown length.

**Comment by TingDaoK** [comment]
Just released a latest version with the fix, [here](https://github.com/awslabs/aws-crt-java/tree/v0.26.1), it should be available in maven in around an hour or two.

**Comment by jonathanswenson** [comment]
The updated version appears to resolve this for me.

**Comment by fcoutinho-uolinc** [comment]
It's working for me aswell

**Comment by TingDaoK** [comment]
awesome, closing the issue as resolved

**Comment by jonathanswenson** [comment]
Perhaps I spoke too soon, the above reproduction is no longer failing for me around the 15 minute mark any longer, but I am still running into problems with longer periods of inactivity in a production env. Trying now to see if I can reproduce this scenario / capture logging.

**Comment by jonathanswenson** [comment]
With some minor tweaks to the above repro (run a few times with a 30 second delay and then wait for longer than the keepalive timeout (72,060s) it fails with the same error. 

After this point, every attempt to use the client fails. 

Trace longs generated using: `Log.initLoggingToFile(Log.LogLevel.Trace, "/tmp/s3-log");`

```
... full "successful" request 
[INFO] [2023-09-13T13:57:13Z] [000000016f4f7000] [http-connection] - 0x1282085a0: Client shutdown completed with error 0 (AWS_ERROR_SUCCESS).
[DEBUG] [2023-09-13T13:57:13Z] [000000016f4f7000] [connection-manager] - id=0x128207530: shutdown received for connection (id=0x1282085a0)
[INFO] [2023-09-13T13:57:13Z] [000000016f4f7000] [connection-manager] - id=0x128207530: manager release, failing 0 pending acquisitions
[DEBUG] [2023-09-13T13:57:13Z] [000000016f4f7000] [connection-manager] - id=0x128207530: snapshot - state=2, idle_connection_count=0, pending_acquire_count=0, pending_settings_count=0, pending_connect_count=0, vended_connection_count=0, open_connection_count=0, ref_count=0
[INFO] [2023-09-13T13:57:13Z] [000000016f4f7000] [connection-manager] - id=0x128207530: Destroying self
[DEBUG] [2023-09-13T13:57:13Z] [000000016f4f7000] [channel-bootstrap] - id=0x600001f66e60: releasing bootstrap reference
[DEBUG] [2023-09-13T13:57:13Z] [000000016f4f7000] [channel] - id=0x13960ff00: destroying channel.
[DEBUG] [2023-09-13T13:57:13Z] [000000016f353000] [task-scheduler] - id=0x139751c70: Scheduling s3_client_process_work_task task for immediate execution
[DEBUG] [2023-09-13T13:57:13Z] [000000016f353000] [task-scheduler] - id=0x139751c70: Running s3_client_process_work_task task with <Running> status
[DEBUG] [2023-09-13T13:57:13Z] [000000016f353000] [S3Client] - id=0x139751aa0 s_s3_client_process_work_default - Moving relevant synced_data into threaded_data.
[DEBUG] [2023-09-13T13:57:13Z] [000000016f353000] [S3Client] - id=0x139751aa0 s_s3_client_process_work_default - Processing any new meta requests.
[DEBUG] [2023-09-13T13:57:13Z] [000000016f353000] [S3Client] - id=0x139751aa0 Updating meta requests.
[DEBUG] [2023-09-13T13:57:13Z] [000000016f353000] [S3Client] - id=0x139751aa0 Updating connections, assigning requests where possible.
[INFO] [2023-09-13T13:57:13Z] [000000016f353000] [S3ClientStats] - id=0x139751aa0 Requests-in-flight(approx/exact):28/0  Requests-preparing:28  Requests-queued:0  Requests-network(get/put/default/total):0/0/0/0  Requests-streaming-waiting:0  Requests-streaming:0  Endpoints(in-table/allocated):0/0
[TRACE] [2023-09-13T13:57:13Z] [000000016f4f7000] [http-connection] - id=0x1282085a0: Destroying connection.
[DEBUG] [2023-09-13T13:57:13Z] [000000016f4f7000] [task-scheduler] - id=0x13960ffd8: Running gather_statistics task with <Canceled> status
[TRACE] [2023-09-13T13:57:13Z] [000000016f4f7000] [channel-bootstrap] - releasing client connection args, args=0x600000b7cc30
[TRACE] [2023-09-13T13:57:13Z] [000000016f4f7000] [channel-bootstrap] - destroying client connection args, args=0x600000b7cc30
[DEBUG] [2023-09-13T13:57:13Z] [000000016f4f7000] [channel-bootstrap] - id=0x600001f66e60: releasing bootstrap reference
[DEBUG] [2023-09-13T13:57:15Z] [000000016f4f7000] [task-scheduler] - id=0x60000174f2f0: Running (null) task with <Running> status
[TRACE] [2023-09-13T13:57:15Z] [000000016f4f7000] [socket] - task_id=0x60000174f2f0: timeout task triggered, evaluating timeouts.
// long period of inactivity
[INFO] [2023-09-13T15:58:14Z] [000000016bb3b000] [S3Client] - id=0x139751aa0 Initiating making of meta request
[DEBUG] [2023-09-13T15:58:14Z] [000000016bb3b000] [S3MetaRequest] - id=0x13976c150 Created new Auto-Ranged Put Meta Request.
[DEBUG] [2023-09-13T15:58:14Z] [000000016bb3b000] [channel-bootstrap] - id=0x600001f66e60: acquiring bootstrap reference
[INFO] [2023-09-13T15:58:14Z] [000000016bb3b000] [connection-manager] - id=0x1397663f0: Successfully created
[DEBUG] [2023-09-13T15:58:14Z] [000000016bb3b000] [S3Endpoint] - id=0x600003a4a400: Created connection manager 0x1397663f0 for endpoint
[INFO] [2023-09-13T15:58:14Z] [000000016bb3b000] [S3Client] - id=0x139751aa0: Created meta request 0x13976c150
[DEBUG] [2023-09-13T15:58:14Z] [000000016f353000] [task-scheduler] - id=0x139751c70: Scheduling s3_client_process_work_task task for immediate execution
[DEBUG] [2023-09-13T15:58:14Z] [000000016f353000] [task-scheduler] - id=0x139751c70: Running s3_client_process_work_task task with <Running> status
[DEBUG] [2023-09-13T15:58:14Z] [000000016f353000] [S3Client] - id=0x139751aa0 s_s3_client_process_work_default - Moving relevant synced_data into threaded_data.
[DEBUG] [2023-09-13T15:58:14Z] [000000016f353000] [S3Client] - id=0x139751aa0 s_s3_client_process_work_default - Processing any new meta requests.
[DEBUG] [2023-09-13T15:58:14Z] [000000016f353000] [S3Client] - id=0x139751aa0 Updating meta requests.
[DEBUG] [2023-09-13T15:58:14Z] [000000016f353000] [S3Client] - id=0x139751aa0 Updating connections, assigning requests where possible.
[INFO] [2023-09-13T15:58:14Z] [000000016f353000] [S3ClientStats] - id=0x139751aa0 Requests-in-flight(approx/exact):28/0  Requests-preparing:28  Requests-queued:0  Requests-network(get/put/default/total):0/0/0/0  Requests-streaming-waiting:0  Requests-streaming:0  Endpoints(in-table/allocated):1/1
```

I have not yet determined if it has to be 2 hours (my keepalive timeout).

**Comment by jonathanswenson** [comment]
Sorry, I must have crossed my wires here. Must have just been in the wrong version still somehow. Forced through the appropriate version and am now unable to repro. 

Thanks!

**Comment by singhbaljit** [comment]
I'm still seeing this issue with v0.27.3. I'm using `BlockingInputStreamAsyncRequestBody`.

**Comment by TingDaoK** [comment]
@singhbaljit Please create a new issue with detailed repo steps or CRT logs, it's hard to track closed issue and without any more info. We are pretty sure we fixed the bug causing this hang, but it's, for sure, possible there are more. So, let's create a new issue to track. Thanks!
